tap_snowflake/client.py

```diff
@@ -3,67 +3,223 @@
 """
 
-import sqlalchemy
+from __future__ import annotations
+
+import os
+from pathlib import Path
+from typing import Iterable, List, Tuple
+from uuid import uuid4
 
 from singer_sdk import SQLConnector, SQLStream
+from singer_sdk.helpers._batch import BaseBatchFileEncoding, BatchConfig
+from singer_sdk.streams.core import REPLICATION_FULL_TABLE, REPLICATION_INCREMENTAL
+from snowflake.sqlalchemy import URL
+from sqlalchemy.orm import load_only
+from sqlalchemy.sql import text
 
 
 class SnowflakeConnector(SQLConnector):
     """Connects to the Snowflake SQL source."""
 
+    @classmethod
     def get_sqlalchemy_url(cls, config: dict) -> str:
         """Concatenate a SQLAlchemy URL for use in connecting to the source."""
-        # TODO: Replace this with a valid connection string for your source:
-        return (
-            f"awsathena+rest://{config['aws_access_key_id']}:"
-            f"{config['aws_secret_access_key']}@athena"
-            f".{config['aws_region']}.amazonaws.com:443/"
-            f"{config['schema_name']}?"
-            f"s3_staging_dir={config['s3_staging_dir']}"
-        )
+        params = {
+            "account": config["account"],
+            "user": config["user"],
+            "password": config["password"],
+        }
 
-    @staticmethod
-    def to_jsonschema_type(sql_type: sqlalchemy.types.TypeEngine) -> dict:
-        """Returns a JSON Schema equivalent for the given SQL type.
+        for option in ["database", "schema", "warehouse", "role"]:
+            if config.get(option):
+                params[option] = config.get(option)
 
-        Developers may optionally add custom logic before calling the default
-        implementation inherited from the base class.
-        """
-        # Optionally, add custom logic before calling the super().
-        # You may delete this method if overrides are not needed.
-        return super().to_jsonschema_type(sql_type)
+        return URL(**params)
 
-    @staticmethod
-    def to_sql_type(jsonschema_type: dict) -> sqlalchemy.types.TypeEngine:
-        """Returns a JSON Schema equivalent for the given SQL type.
 
-        Developers may optionally add custom logic before calling the default
-        implementation inherited from the base class.
-        """
-        # Optionally, add custom logic before calling the super().
-        # You may delete this method if overrides are not needed.
-        return super().to_sql_type(jsonschema_type)
+class SelectableSQLStream(SQLStream):
+    """Patches the parent classes get_records method to support stream property selection."""
 
+    def get_selected_columns(self, table) -> List:
+        """Filter column list according to selection criteria."""
+        return [
+            col.name for col in table.columns if self.mask[("properties", col.name)]
+        ]
 
-class SnowflakeStream(SQLStream):
+    def get_records(self, context: dict | None) -> Iterable[dict[str, Any]]:
```

Comment:
@kgpayne - Do you know what it would take upstream in the SDK to remove this method override here for snowflake? I would love to have `get_records()` implement automatically for SQL taps in a generic/robust fashion - and likewise for `process_record()`/`process_*()` for targets.

Reply:
This was overridden to implement selection, with the only additions being the `get_selected_columns` method above, and the snippet below to update the query instance to apply that selection. Everything else was copied verbatim, so should be very easy to port. In our 1:1 we talked about implementing selection here, so as not to block `tap-snowflake` but with a view to port to the SDK in a following iteration 🙂👍 Ditto for incremental.

```python
selected_columns = self.get_selected_columns(table=table)
query = query.options(load_only(*selected_columns))
```